[how the fuck do i fix this bug please to help] overhaul: ling verbs to action buttons

PKPenguin321 opened the pull request:
:cl: PKPenguin321, Armhulen, epochayur
refactor: Changeling powers are now action buttons rather than verbs.
imageadd: Thanks to epochayur for creating and compiling sprites for every single changeling action button!
imageadd: Thanks to Armhulen for the changeling button background sprite!
/:cl:

Still very messy at this stage but fully functional and compiles. Was even able to playtest it and it works as intended.

# **Cant be finished because theres a really stupid bug where if you change bodies (cloning, last resort, monkeying) you lose all your powers. Cant figure out a fix please HELP**

Things that still need doing:
- [x]  Clean up comments and old code
- [x] Move the rest of the verbs to buttons (the systems are all in place but currently I've only moved transformation sting and mimic voice for testing reasons). Probably the most tedious step
- [x] Get sprites for every power that needs a button (passives won't need a button, stings can just use the sting icon). Not technically required but much needed polish.
- [ ] This will assuredly need playtesting, maybe live test merges. (It got one once before and found the bug described above)

PKPenguin321 committed:
ling verbs to action buttons

initial overhaul changes

PKPenguin321 commented:
If you know who made the bg_changeling icon please let me know so I can credit them. I stole it from arm's thingling PR. Will also credit anybody who gives sprites

Armhulen commented:
It was me muahahahaa

Armhulen commented:
Thanks for doing this for me btw, I was about to do the exact same deal

uraniummeltdown commented:
based pk

Kromgar commented:
Will this be a seperate action bar button like bloodcult?

PKPenguin321 commented:
I wasn't aware a second bar was possible, if you can point me to relevant code and it's not too hard to implement I can see what I can do

Kromgar commented:
@PKPenguin321 Cultcode the action buttons for spells/runes.  @Robustin  might be able to help? Also juggernauts, harvesters, and i think artificers use it too

ma44 commented:
@PKPenguin321 When a action bar gets too long, additional action buttons will be added beneath it.

PKPenguin321 commented:
I know but I don't think that's what he means

PKPenguin321 committed:
more powers moved over

bits of cleaning here and there as well

Kromgar commented:
Cult has an additional  action bar seperate from the one in the top left

PKPenguin321 committed:
merges stasis and regenerate, touches up hivemind
also pushes some more powers to buttons
has_sting now returns the actual sting in question so you can do things to it should it exist

PKPenguin321 committed:
fixes up human form
it was used in more places than just humanform.dm
starting to think i should just write out purchasedpowers but it'd be such a pain in the ass

PKPenguin321 committed:
more powers

PKPenguin321 committed:
removes obsoleted files

PKPenguin321 commented:
this won't compile until every power is transferred over, right now i can compile locally and test by keeping the old changling_powers.dm in the .dme though (kind of a hack but it's the only way to test before i'm absolutely finished without getting errors on compile)

MrDoomBringer commented:
@Kromgar, the cult icons to appear at the bottom of the screen, but I believe that's an error, as if you reset button positions they all line up nicely at the top left.

Potato-Masher commented:
The blood cult button locations are intentional.

Potato-Masher commented:
Also:

>inb4 having three rows of action buttons because of all the damn abilities lings have at round start

PKPenguin321 commented:
You can drag buttons around to reposition them you know

epochayur commented:
pkp you dumb slut, reply to me on discord

are the powers listed under '/code/modules/antagonists/changeling/powers/' the only ones that need icons? if so I'll sprite those up

PKPenguin321 commented:
Yeah but a few of those files have more than one power in them. [The list on the wiki's page for changelings is more accurate](https://tgstation13.org/wiki/Changeling), but that one is missing Hivemind Upload/Download and Absorb.

Armhulen commented:
@PKPenguin321

Incase you missed it, I made the background sprite

PKPenguin321 commented:
Yeah I saw, just havent updated the 🆑 yet

PKPenguin321 committed:
Last of the powers

PKPenguin321 commented:
Up next: Fixing compile errors

PKPenguin321 committed:
0 errors, 0 warnings

PKPenguin321 commented:
Only thing that needs changing from a code perspective now is probably the cellular emporium formatting and stylizing, which is a bit fucked up. Also will definitely need a test merge since I'm not 100% confident it'll be bug free.

PKPenguin321 committed:
undo stats_id

turns out im gay

PKPenguin321 committed:
purges obsolete var

PKPenguin321 commented:
well the cellular emporium works now but upon testing i now see that some abilities just don't like getting removed on respec which is a pretty big problem. will test with it some more and keep working...

PKPenguin321 committed:
the answer is parent calls

PKPenguin321 commented:
have u ever felt retarded

PKPenguin321 committed:
cull some comments and dead code

PKPenguin321 commented:
ok. code side is basically done, aside from needing review (and having to add lines for every button that add icons once i get real sprites).
once i get sprites it will be much much prettier, and after that it should probably be tested, but once that's finished it should be good to merge! phew

MoreRobustThanYou commented on code/modules/antagonists/changeling/changeling.dm:
```diff
@@ -28,17 +28,17 @@
+	var/purchasedpowers = list()//aka `stings`
 	var/mimicing = ""
-	var/canrespec = 0
+	var/canrespec = 0//set to 1 in absorb.dm
```
FALSE

MoreRobustThanYou commented on code/modules/antagonists/changeling/changeling.dm:
```diff
@@ -98,14 +99,12 @@
+		for(var/datum/action/changeling/p in purchasedpowers)
 			purchasedpowers -= p
-			p.on_refund(owner.current)
+			p.Remove(owner.current)//Remove() called
```
i don't think the comment is needed

MoreRobustThanYou commented on code/modules/antagonists/changeling/changeling.dm:
```diff
@@ -98,14 +99,12 @@
 	mimicing = ""
 
-/datum/antagonist/changeling/proc/remove_changeling_powers()
+/datum/antagonist/changeling/proc/remove_changeling_powers()//Remove() is ran in this
```
comment not needed

MoreRobustThanYou commented on code/modules/antagonists/changeling/changeling.dm:
```diff
@@ -117,24 +116,24 @@
-		var/obj/effect/proc_holder/changeling/S = new path()
-		if(!S.dna_cost)
+		var/datum/action/changeling/S = new path()
+		if(!S.dna_cost)//autopurchase 0 cost powers
```
why is your code full of obvious comments

MoreRobustThanYou commented on code/modules/antagonists/changeling/changeling.dm:
```diff
@@ -164,8 +163,8 @@
 	geneticpoints -= thepower.dna_cost
-	purchasedpowers += thepower
-	thepower.on_purchase(owner.current)
+	purchasedpowers += thepower//maybe New() the new button here?
```
stop with the dumb comments

PKPenguin321 replied:
my b a lot of these were for self reference to get a grasp on existing ling code and some were TODOs that i forgot to put a TODO on

PKPenguin321 committed:
cleans more comments, uses a define

PKPenguin321 commented:
Hm. Just noticed that I accidentally cut out the powers listing their cost. Will readd at the same time I get sprites since I'll need to comb through every button again

lzimann commented on code/modules/antagonists/changeling/powers/hivemind.dm:
```diff
@@ -1,33 +1,47 @@
+	var/datum/antagonist/changeling/changeling = user.mind.has_antag_datum(/datum/antagonist/changeling)
+	if(changeling.changeling_speak)
+		changeling.changeling_speak = FALSE
+	for(var/datum/action/changeling/p in changeling.purchasedpowers)
```
typeless loop and only assigned the var if the next check is true

lzimann commented on code/modules/antagonists/changeling/changeling.dm:
```diff
@@ -117,24 +115,24 @@
-/datum/antagonist/changeling/proc/has_sting(obj/effect/proc_holder/changeling/power)
-	for(var/obj/effect/proc_holder/changeling/P in purchasedpowers)
+/datum/antagonist/changeling/proc/has_sting(datum/action/changeling/power)
+	for(var/datum/action/changeling/P in purchasedpowers)
```
typeless loop

PKPenguin321 replied:
What does this mean

lzimann replied:
basically:
```
for(var/power in purchasedpowers)
    var/datum/action/changeling/P = power
    ....
```

everything in this list is already this type, so you don't need to do a type check(which is what this kind of loop does)

PKPenguin321 replied:
appreciate it

MoreRobustThanYou replied:
Microoptimizations

lzimann commented on code/modules/antagonists/changeling/changeling.dm:
```diff
@@ -117,24 +115,24 @@
+	for(var/datum/action/changeling/P in purchasedpowers)
 		if(initial(power.name) == P.name)
-			return TRUE
+			return P//I think this should still work
```
keep this `TRUE`, no real reason to return the instance here

lzimann reviewed (changes requested):
this looks fine except a few minor things

PKPenguin321 committed:
that oughtta do it
fixes some typeless loops, undoes a stupid change

epochayur commented:
https://mega.nz/#!I48XFCZT!hi94DDSlAMatPmNGLZRR5rx7ljQKcu64KGX8ZyGhAIg wew okay, here are the sprites, apologies again for taking so long

These will definitely need to be renamed, and I realize the quality of the original sprites may be a little shoddy, but it's better than nothing until someone can come along and change them. A decent amount of these are lifted from the changeling items in items_and_weapons.dmi . Ping me if you need any changes

PKPenguin321 commented:
Thanks a ton. Only one I noticed was missing is Biodegrade, the one that lets you melt cuffs

PKPenguin321 committed:
adds sprites for all but biodegrade

also updates descs of powers with chemical costs

PKPenguin321 commented:
hold that thought just found a bug that makes reviving stasis not work at all

epochayur commented:
https://mega.nz/#!B9N3gCqZ!UDiujVxLaRDIdGdUUcI_LvScjpz6yuxWRhxlKepBslY new link, includes biodegrade and hivemind_link, just use whichever for hivemind you think looks better

PKPenguin321 committed:
adds biodegrade sprite

updates hivemind link sprite

PKPenguin321 commented:
sprites are excellento now. thanks so much again

and if i wasnt clear with my last comment please hold off on testmerging this until i fix this bug, not sure how i didn't catch it before

PKPenguin321 committed:
fixes bad icon state

PKPenguin321 committed:
fixes the revive bug

PKPenguin321 commented on code/modules/antagonists/changeling/powers/fakedeath.dm:
```diff
@@ -19,14 +50,19 @@
+			to_chat(user, "<span class='notice'>Click the Regenerative Stasis button again to revive.</span>")
+			name = "Revive"
+			desc = "We arise once more."
+			button_icon_state = "revive"
```
the name, desc, and button_icon_state changes aren't reflected on the button itself. everything else works but this. not sure why.

YPOQ replied:
These vars are on the action datum, call UpdateButtonIcon() to copy them over to the button.

PKPenguin321 committed:
revive button final fixes

works flawlessly now

PKPenguin321 commented:
wait what it failed

PKPenguin321 commented on code/modules/antagonists/changeling/powers/adrenaline.dm:
```diff
@@ -1,14 +1,15 @@
-	desc = "We evolve additional sacs of adrenaline throughout our body."
+	desc = "We evolve additional sacs of adrenaline throughout our body. Costs 30 chemicals."
 	helptext = "Removes all stuns instantly and adds a short-term reduction in further stuns. Can be used while unconscious. Continued use poisons the body."
+	button_icon_state = "adrenaline"//TODO: Swap out placeholder
```
Bad TODO

PKPenguin321 commented on code/modules/antagonists/changeling/powers/fakedeath.dm:
```diff
@@ -1,13 +1,46 @@
+	if(revive_ready)
+		if((user.stat != DEAD) && !(user.has_trait(TRAIT_FAKEDEATH)))//if they get cloned or something dont give them a free full heal
+			revive_ready = FALSE
+			name = "Reviving Stasis - We fall into a stasis, allowing us to regenerate and trick our enemies."
```
Forgot to update desc and icon here as well

PKPenguin321 commented on code/modules/antagonists/changeling/powers/fakedeath.dm:
```diff
@@ -1,13 +1,46 @@
 	dna_cost = 0
 	req_dna = 1
 	req_stat = DEAD
+	ignores_fakedeath = FALSE
```
now that I think about it this has other checks meaning this can probably always be true

PKPenguin321 committed:
touchups

PKPenguin321 commented:
BUGS FOUND:
- Transforming to a monkey does not properly grant the human form button and removes all of your buttons. Rehumanizing does not give them back
- Dying and being cloned (which puts you into a new body) causes you to lose all of your buttons

PKPenguin321 commented:
Sorry I haven't touched this in so long, been a bit low on free time lately. Going to look at this and see if there's anything I can do to fix it up.

As far as I could tell, old code for verbs had the verbs list being updated on Stat() which seemed like an awful hack so I removed it, and I suspect that this has something to do with why powers aren't being carried over correctly on transform. My plan right now is to I think make a proc that re-grants all purchased powers which I can run whenever the mind with the antag datum enters a new body, but I'm not sure if there's any proc that's ran when a mind enters a new body.

PKPenguin321 commented:
Also note to self I should test what happens if a non-ling mindswaps into a ling

PKPenguin321 committed:
fix attempt

code\modules\antagonists\changeling\changeling.dm:127:warning: : operation has no effect here

PKPenguin321 commented:
my last commit will most likely fail travis. i dont understand the warning the compiler is throwing but playtesting it shows that my solution doesnt even work except for the cellular emporium button and i have no fuckin clue why that might be.
i just dont get it

PKPenguin321 committed:
minor feedback

optimumtact commented:
travis

PKPenguin321 committed:
erge branch 'master' of https://github.com/PKPenguin321/-tg-station into lingbuttons

PKPenguin321 commented:
no merge conflicts on pull?

also i cant fix this bug i dont know how and im out of steam on reading all the fuckign transformation code to figure out how.

PKPenguin321 commented:
i dont know what the compiler warning on my attempted fix means either. either way the fix didnt work so i guess i can just remove it so at least travis passes.

PKPenguin321 committed:
comments out travis-killing code

optimumtact commented on code/modules/antagonists/changeling/changeling.dm:
```diff
@@ -129,9 +129,11 @@
-	for(var/path in purchasedpowers)
-		var/datum/action/changeling/S = new path()
-			S.Grant(owner.current)
+//	for(var/path in purchasedpowers)
```
you know this is required code right?

PKPenguin321 replied:
Yes it kills travis see the comments below it
It does nothing in testing also i wrote it so yes i know what it is

SpaceManiac commented:
You wrote:
```dm
	for(var/path in purchasedpowers)
		var/datum/action/changeling/S = new path()
			S.Grant(owner.current)
```
You probably meant:
```dm
	for(var/path in purchasedpowers)
		var/datum/action/changeling/S = new path()
		S.Grant(owner.current)
```

PKPenguin321 commented:
holy heck you might be right

PKPenguin321 commented:
ok so it still didnt work, only the cellular emporium button is granted by regain_powers and the rest of the buttons are still lost. at least it passes travis now without commenting it out. will try putting it on mob login.

PKPenguin321 committed:
why the fuck dont my code work

PKPenguin321 commented:
issue persists even when the proc is ran on mob login 😠 😢

PKPenguin321 commented:
why the fuck my code dont work
![why](https://user-images.githubusercontent.com/8345184/37365735-96733d92-26bb-11e8-95ce-5ec18733f89d.png)

KorPhaeron commented:
I was excited for this PR but it is pretty clearly dead, please finish it when you get the chance

PKPenguin321 commented:
basically action buttons seem to poof out of existence and DO NOT want to be regranted if you change mob for some reason. i suspect it's a bug with action buttons but am not about to read all the action button code to figure it out, just too exhausted with it. rip
